### PR TITLE
Added ability to encrypt decrypt and hash conditionally

### LINKED ISSUE #1269: Add ability to encrypt, decrypt and hash fields conditionally
Should be pretty straightforward. We have already done something similar for `force` and `remove` rules.

> Might require changes in security rule builder.

>If you want this feature to be implemented, give it a thumbs up reaction, so that we can determine which features are important to you.
>👍

### PR DESCRIPTION
**Pull request by KavishShah09**
Fixes #1269 
Also removed a lot of warnings about redundant structs being used in tests

### CODE REVIEW

File: gateway/modules/auth/match_test.go [hunk 1]
```diff
@@ -334,19 +334,34 @@ func TestModule_matchEncrypt(t *testing.T) {
...
+		{
+			name: "rule clause - deny",
+			m:    &Module{project: "project"},
+			args: args{
+				rule: &config.Rule{Rule: "encrypt", Clause: &config.Rule{Rule: "deny"}},
+			},
+		},
 	}
 
 	for _, tt := range tests {
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:362]: You think we should check if the fields are getting encrypted or not? A simple check to see whether its changed or not can be helpful

File: gateway/modules/auth/match_test.go [hunk 2]
```diff
@@ -392,19 +407,34 @@ func TestModule_matchDecrypt(t *testing.T) {
...
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:435]: We should check if the value was actually decrypted or not

File: gateway/modules/auth/match_test.go [hunk 3]
```diff
@@ -481,11 +512,26 @@ func Test_matchHash(t *testing.T) {
...
 		t.Run(tt.name, func(t *testing.T) {
-			got, err := matchHash(context.Background(), tt.args.rule, tt.args.args)
+			got, err := tt.m.matchHash(context.Background(), "project", tt.args.rule, tt.args.args, make(map[string]interface{}))
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:534]: We should check if the value was actually hashed or not

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: gateway/modules/auth/match_test.go (hunk 1 above)
Comment (KavishShah09) [reply gateway/modules/auth/match_test.go:362]: Sure, on it!

File: gateway/modules/auth/match_test.go (hunk 3 above)
Comment (KavishShah09) [reply gateway/modules/auth/match_test.go:534]: 👍

File: gateway/modules/auth/match_test.go (hunk 2 above)
Comment (KavishShah09) [reply gateway/modules/auth/match_test.go:435]: 👍

**Review by sharadregoti — APPROVED** [review]

File: gateway/modules/auth/match_test.go [hunk 4]
```diff
@@ -330,27 +332,59 @@ func TestModule_matchEncrypt(t *testing.T) {
...
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			got, err := tt.m.matchEncrypt(context.Background(), tt.args.rule, tt.args.args)
+			got, err := tt.m.matchEncrypt(context.Background(), "project", tt.args.rule, tt.args.args, make(map[string]interface{}))
 			if (err != nil) != tt.wantErr {
 				t.Errorf("Module.matchEncrypt() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
+			if tt.shouldChange && reflect.DeepEqual(tt.originalArgs, tt.args.args) {
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:385]:
```suggestion
			if tt.shouldChange == reflect.DeepEqual(tt.originalArgs, tt.args.args) {
```

File: gateway/modules/auth/match_test.go [hunk 5]
```diff
@@ -330,27 +332,59 @@ func TestModule_matchEncrypt(t *testing.T) {
...
+				t.Errorf("Module.matchEncrypt() = %v, wanted: change from  %v", tt.args.args, tt.originalArgs)
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:386]:
```suggestion
				t.Errorf("Module.matchEncrypt() = %v, wanted change - %v", tt.args.args, tt.shouldChange)
```

File: gateway/modules/auth/match_test.go [hunk 6]
```diff
@@ -388,27 +424,59 @@ func TestModule_matchDecrypt(t *testing.T) {
...
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			got, err := tt.m.matchDecrypt(context.Background(), tt.args.rule, tt.args.args)
+			got, err := tt.m.matchDecrypt(context.Background(), "project", tt.args.rule, tt.args.args, make(map[string]interface{}))
 			if (err != nil) != tt.wantErr {
 				t.Errorf("Module.matchDecrypt() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
+			if tt.shouldChange && reflect.DeepEqual(tt.originalArgs, tt.args.args) {
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:477]:
```suggestion
			if tt.shouldChange == reflect.DeepEqual(tt.originalArgs, tt.args.args) {
```

File: gateway/modules/auth/match_test.go [hunk 7]
```diff
@@ -388,27 +424,59 @@ func TestModule_matchDecrypt(t *testing.T) {
...
+				t.Errorf("Module.matchEncrypt() = %v, wanted: change from  %v", tt.args.args, tt.originalArgs)
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:478]:
```suggestion
				t.Errorf("Module.matchEncrypt() = %v, wanted change - %v", tt.args.args, tt.shouldChange)
```

File: gateway/modules/auth/match_test.go [hunk 8]
```diff
@@ -481,15 +552,40 @@ func Test_matchHash(t *testing.T) {
...
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			got, err := matchHash(context.Background(), tt.args.rule, tt.args.args)
+			got, err := tt.m.matchHash(context.Background(), "project", tt.args.rule, tt.args.args, make(map[string]interface{}))
 			if (err != nil) != tt.wantErr {
 				t.Errorf("matchHash() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
+			if tt.shouldChange && reflect.DeepEqual(tt.originalArgs, tt.args.args) {
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:586]:
```suggestion
			if tt.shouldChange == reflect.DeepEqual(tt.originalArgs, tt.args.args) {
```

File: gateway/modules/auth/match_test.go [hunk 9]
```diff
@@ -481,15 +552,40 @@ func Test_matchHash(t *testing.T) {
...
+				t.Errorf("Module.matchEncrypt() = %v, wanted: change from  %v", tt.args.args, tt.originalArgs)
```
Comment (YourTechBud) [inline gateway/modules/auth/match_test.go:587]:
```suggestion
				t.Errorf("Module.matchEncrypt() = %v, wanted change - %v", tt.args.args, tt.shouldChange)
```

**Review by YourTechBud — CHANGES_REQUESTED** [review]

**Pull request closed without merging**